crates/bevy_ecs/src/query/iter.rs

```diff
@@ -1116,61 +1116,57 @@ impl<'w, 's, D: QueryData, F: QueryFilter, I: Iterator<Item: Borrow<Entity>>>
     }
 
     /// Safety:
+    /// All arguments must stem from the same valid `QueryManyIter`.
+    ///
     /// The lifetime here is not restrictive enough for Fetch with &mut access,
     /// as calling `fetch_next_aliased_unchecked` multiple times can produce multiple
     /// references to the same component, leading to unique reference aliasing.
     ///
     /// It is always safe for shared access.
     #[inline(always)]
-    unsafe fn fetch_next_aliased_unchecked(&mut self) -> Option<D::Item<'w>> {
-        for entity in self.entity_iter.by_ref() {
+    unsafe fn fetch_next_aliased_unchecked(
+        entity_iter: impl Iterator<Item: Borrow<Entity>>,
```

Comment:
It's unfortunate to have to pass all of these parameters separately.  Would it make sense to wrap everything but `entity_iter` into a new `struct`?  

You could extract the body of the loop to a method on that new struct that takes `&mut self` and `Entity` and returns an `Option<D::Item>`.  

That could also simplify your `QueryManyUniquIter` PR slightly, since you could move the whole struct when changing the `entity_iter` type instead of having to list every field.

Reply:
That is true, I think I'll try and extract these extra parameters into a `QueryManyIterState`.

I think this will regress struct size for some cases (Lead to a size doubling in the worst case?), but it might be worth the simplification.

Reply:
On second thought, I think it'd be better to leave it like this. Extracting the extra parameters into a separate struct only works if that struct is a field of `QueryManyIter`, because we otherwise run into missing lifetime extension.
This would cleanup the fetch/next methods, but split the struct definition while also regressing the layout.